### PR TITLE
iox-#1755 Reintroduce lazy evaluation for the logging

### PR DESCRIPTION
**Pull request by elBoberido**
## Pre-Review Checklist for the PR Author

1. [x] Code follows the coding style of [CONTRIBUTING.md][contributing]
1. [x] Tests follow the [best practice for testing][testing]
1. [x] Changelog updated [in the unreleased section][changelog] including API breaking changes
1. [x] Branch follows the naming format (`iox-123-this-is-a-branch`)
1. [x] Commits messages are according to this [guideline][commit-guidelines]
    - [x] Commit messages have the issue ID (`iox-#123 commit text`)
    - [x] Commit author matches [Eclipse Contributor Agreement][eca] (and ECA is signed)
1. [x] Update the PR title
   - Follow the same conventions as for commit messages
   - Link to the relevant issue
1. [x] Relevant issues are linked
1. [x] Add sensible notes for the reviewer
1. [x] All checks have passed (except `task-list-completed`)
1. [x] All touched (C/C++) source code files from `iceoryx_hoofs` are added to `./clang-tidy-diff-scans.txt`
1. [x] Assign PR to reviewer

[commit-guidelines]: https://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html
[eca]: http://www.eclipse.org/legal/ECA.php
[contributing]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/CONTRIBUTING.md#coding-style
[testing]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/doc/website/concepts/best-practice-for-testing.md
[changelog]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/doc/website/release-notes/iceoryx-unreleased.md

## Notes for Reviewer

This PR introduces a new macro for logging which utilizes lazy evaluation. The port to the new macro will happen in a follow up PR.

## Checklist for the PR Reviewer

- [x] Commits are properly organized and messages are according to the guideline
- [x] Code according to our coding style and naming conventions
- [x] Unit tests have been written for new behavior
- [x] Public API changes are documented via doxygen
- [x] Copyright owner are updated in the changed files
- [x] All touched (C/C++) source code files from `iceoryx_hoofs` have been added to `./clang-tidy-diff-scans.txt`
- [x] PR title describes the changes

## Post-review Checklist for the PR Author

1. [x] All open points are addressed and tracked via issues

## References

- Relates to #1755

### CODE REVIEW

**Comment by elBoberido** [comment]
@FerdinandSpitzschnueffler  @mossmaurice @dkroenke @MatthiasKillat @elfenpiff feel free to add yourself as reviewer

File: doc/design/logging.md
Comment (FerdinandSpitzschnueffler) [inline doc/design/logging.md:100]:
```suggestion
The `IOX_LOG` macro is intended for general use. It has two parameters. The first one
```

File: doc/design/logging.md
Comment (FerdinandSpitzschnueffler) [inline doc/design/logging.md:104]:
```suggestion
the items to log and multiple items can be logged by using the '<<' operator.
```

File: iceoryx_hoofs/include/iceoryx_hoofs/error_reporting/error_logging.hpp
Comment (FerdinandSpitzschnueffler) [inline iceoryx_hoofs/include/iceoryx_hoofs/error_reporting/error_logging.hpp:28]: Please add a `@param` for `msg_stream` here and for the the following macros.

**Review by FerdinandSpitzschnueffler — COMMENTED** [review]
LGTM, only nitpicks.

**Review by FerdinandSpitzschnueffler — APPROVED** [review]

**Comment by elBoberido** [comment]
@mossmaurice @dkroenke @MatthiasKillat @elfenpiff anyone who would like to do the second review?

We desperately need to change the number of reviewers. There is another open PR which is based on this one and I have a further branch which is ready to review. With this speed I won't be able to get this in this year and I'm currently blocked by the slow review process.

**Comment by elBoberido** [comment]
@mossmaurice @dkroenke @MatthiasKillat @elfenpiff anyone?

File: tools/scripts/clang_tidy_check.sh
Comment (mossmaurice) [inline tools/scripts/clang_tidy_check.sh:25]: :heart:

**Review by mossmaurice — APPROVED** [review]
LGTM, thanks!

**Pull request merged**
